### ISSUE TITLE
Diagnostics for using ctor of typedef could be better

### ISSUE THREAD

**Issue by Manishearth** [rust-lang/rust#36583]
```

type Bar = foo::Foo;
fn main() {
    let x: Bar = Bar(1);
}

pub mod foo {
 pub struct Foo(u8);
}
```

([playpen](https://is.gd/tWPL2D))

This gives the error

```
error: unresolved name `Bar` [--explain E0425]
 --> <anon>:4:18
  |>
4 |>     let x: Bar = Bar(1);
  |>                  ^^^

error: aborting due to previous error
```

This happens because `type`defs only create a name in "type" space, whereas struct ctors are in "value" space. One way to fix this is to use a reexport instead of a typedef. (If you don't control that code, you should import the actual path)

We should detect this situation and print out some suggestions.

**Comment by petrochenkov** [comment]
FWIW, I (sometimes) work on a patch, that includes resolving all unresolved names in the alternative namespace.
With it the error message will look like `expected function, found type alias Bar`. No suggestions, but still an improvement compared to just `unresolved name`.

**Comment by petrochenkov** [comment]
I don't like how resolution-related code is full of randomly added ad hoc suggestions and custom error messages. I'd like to have the error messages unified in one "expected-found" form, and to find a way to gather all "did you mean *this*?" messages in some kind of table in which suggestions are functions of the lookup context and the found entity `Suggestion = F(where_we_search, what_is_found)`.

**Comment by Manishearth** [comment]
Works for me!

Yeah, when I was last hacking on the resolution system I realized how hard it was to re-search for the same name.

**Comment by Manishearth** [comment]
(might not get time to work on this after all. may pick up later)

**Comment by petrochenkov** [comment]
Fixed in 1.16.0
```
rustc 1.16.0-beta.2 (bc15d5281 2017-02-16)
error[E0423]: expected function, found type alias `Bar`
 --> <anon>:3:18
  |
3 |     let x: Bar = Bar(1);
  |                  ^^^ did you mean `Bar { /* fields */ }`?

error: aborting due to previous error
```
